### PR TITLE
[ENH] Cinn based forecaster

### LINKED ISSUE #4924: [ENH] Normalizing Flow/cINN based forecaster from the BigDEAL competition
**Is your feature request related to a problem? Please describe.**
It would be cool if we implement and support some Normalizing Flows for forecasting. In the BigDEAL challenge, the second-placed methods use such an approach.

**Describe the solution you'd like**
A new forecaster that implements this solution. Since the solution consists of different subcomponents, it would be useful if this model uses the #4652 internally.

**Describe alternatives you've considered**
Implementing it without using a graphpipeline. Probably would reduce the reusability of subcomponents.

**Additional context**
Requires an additional soft-dependency. E.g. FrEIA.

### PR DESCRIPTION
**Pull request by benHeid**
## Note relies on 
* DeepLearning base design is merged 
* CurveFitForecaster is merged
* Merger #5351 

#### Reference Issues/PRs
Fixes #4924

#### What does this implement/fix? Explain your changes.
It implements the method of the BigDEAL Challenge that won the peak shape and the peak timing forecasting tasks.

Furthermore, it adapts a bit the base classes for torch forecasters to make them more flexible.

#### Does your contribution introduce a new dependency? If yes, which one?
It introduces FrEIA as a dependency. This is a framework for implementing reversible neural network architectures. 

##### For all contributions
- [x] I've added myself to the [list of contributors](https://github.com/sktime/sktime/blob/main/CONTRIBUTORS.md) with any new badges I've earned :-)
  How to: add yourself to the [all-contributors file](https://github.com/sktime/sktime/blob/main/.all-contributorsrc) in the `sktime` root directory (not the `CONTRIBUTORS.md`). Common badges: `code` - fixing a bug, or adding code logic. `doc` - writing or improving documentation or docstrings. `bug` - reporting or diagnosing a bug (get this plus `code` if you also fixed the bug in the PR).`maintenance` - CI, test framework, release.
  See here for [full badge reference](https://allcontributors.org/docs/en/emoji-key)
- [x] Optionally, I've added myself and possibly others to the [CODEOWNERS](https://github.com/sktime/sktime/blob/main/CODEOWNERS) file - do this if you want to become the owner or maintainer of an estimator you added.
  See here for further details on the [algorithm maintainer role](https://www.sktime.net/en/latest/get_involved/governance.html#algorithm-maintainers).
- [x] The PR title starts with either [ENH], [MNT], [DOC], or [BUG]. [BUG] - bugfix, [MNT] - CI, test framework, [ENH] - adding or improving code, [DOC] - writing or improving documentation or docstrings.

##### For new estimators
- [x] I've added the estimator to the API reference - in `docs/source/api_reference/taskname.rst`, follow the pattern.
- [x] I've added one or more illustrative usage examples to the docstring, in a pydocstyle compliant `Examples` section.
- [x] If the estimator relies on a soft dependency, I've set the `python_dependencies` tag and ensured
  dependency isolation, see the [estimator dependencies guide](https://www.sktime.net/en/latest/developer_guide/dependencies.html#adding-a-soft-dependency).

### CODE REVIEW

**Comment by benHeid** [comment]
# TODOs:
* [ ] Add reference to paper
* [ ] Do some refactoring
* [ ] Fix soft dependency issues.

**Comment by fkiraly** [comment]
@yarnabrina, highlighting an observation: the new CI elements do not test the new forecaster, but the old CI does.

As far as I understood the setup, the new CI elements *should* test the forecaster though - so it looks like that it's not happening here might be a logic bug in the new CI elements.

**Comment by fkiraly** [comment]
ready for review?

File: docs/source/api_reference/forecasting.rst [hunk 1]
```diff
@@ -341,6 +341,17 @@ Transformer (deep learning) based forecasters
     LTSFDLinearForecaster
     LTSFNLinearForecaster
 
+Deep generative model based forecasters
```
Comment (fkiraly) [inline docs/source/api_reference/forecasting.rst:344]: should this be a new category with single inhabitant, or should we merge categories?

File: pyproject.toml [hunk 2]
```diff
@@ -120,6 +122,7 @@ all_extras_pandas2 = [
   "dtw-python",
   'esig==0.9.7; python_version < "3.10"',
   'filterpy>=1.4.5; python_version < "3.11"',
+  'FrEIA; python_version < "3.12"',
```
Comment (fkiraly) [inline pyproject.toml:125]: I think python package names ignore capitalization

**Comment by benHeid** [comment]
> ready for review?

Yes

File: sktime/forecasting/base/adapters/_pytorch.py [hunk 3]
```diff
@@ -56,8 +63,6 @@ def _fit(self, y, fh, X=None):
         X : iterable-style or map-style dataset
             see (https://pytorch.org/docs/stable/data.html) for more information
         """
-        from sktime.forecasting.base import ForecastingHorizon
```
Comment (fkiraly) [inline sktime/forecasting/base/adapters/_pytorch.py:59]: it might be safer to have it in here to avoid cirular imports?

File: sktime/forecasting/base/adapters/_pytorch.py [hunk 4]
```diff
@@ -66,36 +71,37 @@ def _fit(self, y, fh, X=None):
             self._fh = fh
 
         self._y = y
-
-        if type(fh) is ForecastingHorizon:
-            self.network = self._build_network(fh._values[-1])
+        if isinstance(self.fh, ForecastingHorizon):
```
Comment (fkiraly) [inline sktime/forecasting/base/adapters/_pytorch.py:74]: seems like a superfluous check, in `_fit`, the `fh` is guaranteed `ForecastingHorizon`

File: sktime/forecasting/base/adapters/_pytorch.py [hunk 5]
```diff
@@ -56,8 +63,6 @@ def _fit(self, y, fh, X=None):
         X : iterable-style or map-style dataset
             see (https://pytorch.org/docs/stable/data.html) for more information
         """
-        from sktime.forecasting.base import ForecastingHorizon
-
```
Comment (fkiraly) [inline sktime/forecasting/base/adapters/_pytorch.py:60]: the saving of `_fh` and `_y` seems problematic, this should be controlled in the base class

File: sktime/forecasting/base/adapters/_pytorch.py [hunk 6]
```diff
@@ -66,36 +71,37 @@ def _fit(self, y, fh, X=None):
...
+
+    def _run_epoch(self, epoch, dataloader):
+        for x, y in dataloader:
+            y_pred = self.network(x)
+            loss = self._criterion(y_pred, y)
+            self._optimizer.zero_grad()
+            loss.backward()
+            self._optimizer.step()
 
+    def _instatiate_optimizer(self):
```
Comment (fkiraly) [inline sktime/forecasting/base/adapters/_pytorch.py:96]: typo - instaNtiate?

**Review by fkiraly — CHANGES_REQUESTED** [review]
Some requests:

* extension of the adapter is ok, since downwards compatible.
* there is some legacy code controlling saving to `self._fh`, `self._y` - that does not look right, it should happen in `fit` (base class), see comments
* classes should be `UpperCamelCase`

Question:
why are save/load methods specific to the new forecaster? Is the pytorch base class not enough for this?

File: sktime/forecasting/base/adapters/_pytorch.py (hunk 3 above)
Comment (benHeid) [reply sktime/forecasting/base/adapters/_pytorch.py:59]: I would be very surprised if something from base.py is importing something directly or indirectly from _pytorch.py

If this would be the case, I think this would indicate some design issues since something from a base module is importing something that is specific to a certain class. Thus, if this would be the case, we should factor the imported component from the specific module out.

File: sktime/forecasting/base/adapters/_pytorch.py (hunk 4 above)
Comment (benHeid) [reply sktime/forecasting/base/adapters/_pytorch.py:74]: You are probably right, I just replaced the old code with isinstance to be more pythonic. I will check if this is needed.

File: docs/source/api_reference/forecasting.rst (hunk 1 above)
Comment (benHeid) [reply docs/source/api_reference/forecasting.rst:344]: We can merge it with the Transformer (deep learning) based forecasters. I would propose instead case to name it deep learning based forecasters. Would probably also fit better tk the LTFSDLinearForecaster, since I am unsure if they are transformers or not.

File: pyproject.toml (hunk 2 above)
Comment (benHeid) [reply pyproject.toml:125]: I took the naming from pypi: https://pypi.org/project/FrEIA/

**Comment by fkiraly** [comment]
failures seem genuine - were save/load needed? if so, what is different from the base network class?

**Comment by benHeid** [comment]
I am unsure and have to check it again. But after looking on the code, I extracted the parameters of the internal cINN and put it in a dict which was then pickled together with the rest but without the network. As far as I remember me correctly, the reason was that pickling the cINN failed or does not work correctly due to functions that were passed to the underlying FrEIA framework, which used them to construct the network structure. 

But I need to check it. (That happens if you are not commenting your own code... 😅)

**Comment by fkiraly** [comment]
So, let's just add the save/load functions back then? It seems they were needed.

File: sktime/forecasting/base/adapters/_pytorch.py (hunk 5 above)
Comment (benHeid) [reply sktime/forecasting/base/adapters/_pytorch.py:59]: Fixed by other PR

File: sktime/forecasting/base/adapters/_pytorch.py (hunk 4 above)
Comment (benHeid) [reply sktime/forecasting/base/adapters/_pytorch.py]: I removed that check

**Comment by fkiraly** [comment]
hmm, it appears there are more faliures than previously - can I help with this?

**Comment by benHeid** [comment]
> hmm, it appears there are more faliures than previously - can I help with this?

I got a bit distracted. And I am a bit confused because `test_predict_time_index_in_sample_full` is executed although the tag: `capability:insample` is false.

**Comment by fkiraly** [comment]
> I got a bit distracted. And I am a bit confused because `test_predict_time_index_in_sample_full` is executed although the tag: `capability:insample` is false.

In that case, `_predict` should return `NotImplementedError` - this is not (yet?) done by the boilerplate.

**Review by fkiraly — APPROVED** [review]
No more failures, so all fixed.

File: docs/source/api_reference/forecasting.rst [hunk 7]
```diff
@@ -350,7 +350,7 @@ Structural time series models
 
     DynamicFactor
 
-Transformer (deep learning) based forecasters
+Deep learning based forecasters
```
Comment (yarnabrina) [inline docs/source/api_reference/forecasting.rst:353]: May I request to not change this? If `neuralforecast` PR (#5962) is approved and merged, it'll have conflicts, but more importantly we'll be able to add more deep learning based forecasters and it may be better to group them into transformer based, RNN based, etc.

File: sktime/forecasting/base/adapters/_pytorch.py (hunk 3 above)
Comment (fkiraly) [reply sktime/forecasting/base/adapters/_pytorch.py:58]: makes sense.

File: docs/source/api_reference/forecasting.rst (hunk 1 above)
Comment (fkiraly) [reply docs/source/api_reference/forecasting.rst]: Yes, makes sense. Perhaps separate PR to avoid clash with @yarnabrina PR?

File: docs/source/api_reference/forecasting.rst (hunk 7 above)
Comment (benHeid) [reply docs/source/api_reference/forecasting.rst:353]: I think you need not to revert the changes. Since If I do understand it correctly, the mentioned LTSF forecaster are not transformers.

**Review by fkiraly — APPROVED** [review]
For the release, I've reverted the user facing dependency sets to their original state, to avoid problematic installations popping up within a version cycle where both `tensorflow` and `torch` are present. Testing coverage should still be ensured via the `dl` dependency set.
